ethcore/src/snapshot/service.rs

```diff
@@ -361,11 +352,27 @@ impl Service {
 
 		// The old database looks like this:
 		// [genesis, best_ancient_block] ... [first_block, best_block]
-		// If we are fully synced neither `best_ancient_block` nor `first_block` is set, and we can assume that the whole range from [genesis, best_block] is imported.
-		// The new database only contains the tip of the chain ([first_block, best_block]),
+		// If we are fully synced neither `best_ancient_block` nor `first_block` is set, and we can
+		// assume that the whole range from [genesis, best_block] is imported.
+		// The new database only contains the tip of the chain ([new_first_block, new_best_block]),
 		// so the useful set of blocks is defined as:
 		// [0 ... min(new.first_block, best_ancient_block or best_block)]
+		//
+		// If, for whatever reason, the old db does not have ancient blocks (i.e.
+		// `best_ancient_block` is `None`) AND a non-zero `first_block`, such that the old db looks
+		// like [old_first_block..old_best_block] (which may or may not partially overlap with
+		// [new_first_block..new_best_block]) we do the conservative thing and do not migrate the
+		// old blocks.
 		let find_range = || -> Option<(H256, H256)> {
+			// TODO: In theory, if the current best_block is > new first_block (i.e. they overlap)
+			//  we could salvage them but what if there's been a re-org at the boundary and the two
+			//  chains do not match anymore? Would it actually work?
```

Comment:
I'd rather drop the old blocks if we are unsure. We should definitely verify if at least couple of latest block we had in the old db matches the new one.

Reply:
For PoW, is there a limit on the length of reorgs?

Reply:
Not really, it's only probabilistic. But usually the limiting factor is the state pruning, I don't remember though what value it's set to.